pages/user.js

```diff
@@ -34,9 +71,125 @@ class User extends Component {
         return { isLoggedIn, userData };
     }
 
+    componentDidMount() {
+        this.handleSocket();
+    }
+
+    onCloseNotification() {
+        const { notificationKey: key } = this.state;
+        this.acceptChallengeRequest();
+        notification.close(key);
+    }
+
+    getUserList() {
+        this.socket.emit(getUser, {});
+    }
+
+    rejectChallengeRequest() {
+        this.socket.emit(rejectChallenge, {});
+    }
+
+    acceptChallengeRequest() {
+        this.socket.emit(acceptChallenge, {});
+    }
+
+    // call this when user want to end current challenge
+    leaveChat() {
+        const { connected, room } = this.state;
+        if (connected) {
+            this.socket.emit(leaveRoom);
+            this.socket.leave(room);
+            this.setState({ room: '' });
+        }
+    }
+
+    handleSocket() {
+        const { connect } = CLIENT_SYSTEM_EVENTS;
+        const { userData: user, isLoggedIn } = this.props;
+        const { room } = this.state;
+        const { _id, name, picture } = user;
+        this.socket = io();
+
+        this.socket.on(connect, () => {
+            this.setState({ connected: true });
+
+            if (isLoggedIn) {
+                this.socket.emit(login, { _id, name, picture });
+            }
+        });
+
+        this.socket.on(users, data => {
+            this.setState({ onlineUsers: data.users });
+            const { onlineUsers } = this.state;
+            console.log(`users called ${onlineUsers.length}`);
+        });
+
+        this.socket.on(challengeStart, data => {
+            this.setState({ room: data.room });
+            // TODO show notification to user; challengeStartNotification(data.user.name);
+        });
+
+        this.socket.on(ackChallengeRequest, data => {
+            const { room: roomId } = data;
+            this.setState({ room: roomId });
+            // TODO show notification to user; challengeRequestNotification(data.user);
+        });
+
+        this.socket.on(challengeRequest, data => {
+            const { room: roomId, user: sender } = data;
+            this.setState({ room: roomId });
+            this.openNotification(sender);
+            // TODO show notification to user; challengeRequestNotification(data.user);
+        });
+
+        this.socket.on(challengeEnd, () => {
+            // TODO show notification to user; challengeEndNotification();
```

Comment:
remove all these separate `TODO`'s, add one todo at the top of `handleSocket` saying that you need to figure out how to `show notification to user`